src/solver/optim/ortoolsImpl/include/antares/solver/optim/ortoolsImpl/linearProblem.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2007-2024, RTE (https://www.rte-france.com)
+ * See AUTHORS.txt
+ * SPDX-License-Identifier: MPL-2.0
+ * This file is part of Antares-Simulator,
+ * Adequacy and Performance assessment for interconnected energy networks.
+ *
+ * Antares_Simulator is free software: you can redistribute it and/or modify
+ * it under the terms of the Mozilla Public Licence 2.0 as published by
+ * the Mozilla Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Antares_Simulator is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Mozilla Public Licence 2.0 for more details.
+ *
+ * You should have received a copy of the Mozilla Public Licence 2.0
+ * along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+ */
+
+#pragma once
+
+#include <antares/solver/optim/api/linearProblem.h>
+#include <antares/solver/optim/ortoolsImpl/mipConstraint.h>
+#include <antares/solver/optim/ortoolsImpl/mipSolution.h>
+#include <antares/solver/optim/ortoolsImpl/mipVariable.h>
+
+// forward declaration
+namespace operations_research
+{
+class MPSolver;
+class MPSolverParameters;
+class MPObjective;
+} // namespace operations_research
+
+namespace Antares::Solver::Optim::OrtoolsImpl
+{
+
+class OrtoolsLinearProblem final: public Api::ILinearProblem
+{
+public:
+    OrtoolsLinearProblem(bool isMip, const std::string& solverName);
+    ~OrtoolsLinearProblem() final = default;
+
+    OrtoolsMipVariable* addNumVariable(double lb, double ub, const std::string& name) override;
+    OrtoolsMipVariable* addIntVariable(double lb, double ub, const std::string& name) override;
+    OrtoolsMipVariable* addVariable(double lb,
+                                    double ub,
+                                    bool integer,
+                                    const std::string& name) override;
+    OrtoolsMipVariable* getVariable(const std::string& name) const override;
+
+    OrtoolsMipConstraint* addConstraint(double lb, double ub, const std::string& name) override;
+    OrtoolsMipConstraint* getConstraint(const std::string& name) const override;
+
+    void setObjectiveCoefficient(Api::IMipVariable* var, double coefficient) override;
+    double getObjectiveCoefficient(const Api::IMipVariable* var) const override;
+
+    void setMinimization() override;
+    void setMaximization() override;
+
+    bool isMinimization() const override;
+    bool isMaximization() const override;
+
+    OrtoolsMipSolution* solve(bool verboseSolver) override;
+
+private:
+    std::unique_ptr<operations_research::MPSolver> mpSolver_;
+    operations_research::MPObjective* objective_;
+    operations_research::MPSolverParameters params_;
+
+    std::map<std::string, std::unique_ptr<OrtoolsMipVariable>> variables_;
+    std::map<std::string, std::unique_ptr<OrtoolsMipConstraint>> constraints_;
```

Comment:
About : 
```c++
std::map<std::string, std::unique_ptr<OrtoolsMipVariable>> variables_;
std::map<std::string, std::unique_ptr<OrtoolsMipConstraint>> constraints_;
```
These 2 data members are private and use class **OrtoolsMipVariable** or **OrtoolsMipConstraint**.

Using **OrtoolsMipVariable** or  **OrtoolsMipConstraint** objects internally (or privately if you like), makes the internal code more complex than it could be by using **ortools** native types.

If we use, **ortools** native types, we don't need these 2 data members  (**variables_** & **constraints_**) : these lists are already stored in **mpSolver_**.

For example, in **OrtoolsLinearProblem::addVariable(...)**, we could have : 
```C++
OrtoolsMipVariable* OrtoolsLinearProblem::addVariable(double lb,
                                                      double ub,
                                                      bool integer,
                                                      const std::string& name)
{
    if (mpSolver_.LookupVariableOrNull(name))
    {
        logs.error() << "This variable already exists: " << name;
        throw ElemAlreadyExists();
    }

    auto* mpVar = mpSolver_->MakeVar(lb, ub, integer, name);

    if (!mpVar)
    {
        logs.error() << "Couldn't add variable to Ortools MPSolver: " << name;
    }

    return std::make_unique<OrtoolsMipVariable>(mpVar).get();
}
```
This function looks simpler like this.

Reply:
```cpp
    return std::make_unique<OrtoolsMipVariable>(mpVar).get();
```
This function would return a dangling pointer 👎🏻

Reply:
ok, but if you agree with the spirit of the remark, we can find another solution.

Reply:
ok.
Are we forced to return a pointer to **OrtoolsMipVariable** here ?
Could we not return a **OrtoolsMipVariable** ? In that case, it would be simple : 
```c++
OrtoolsMipVariable OrtoolsLinearProblem::addVariable(...)
{
    ....
    return mpVar;
}
```

Reply:
```cpp
auto var_to_return = std::make_unique<OrtoolsMipVariable>(mpVar);
return var_to_return.get();
``` 
is completely equivalent, since `var_to_return` is destructed at the end of scope.

Reply:
> Covariance doesn't apply here since e.g `OrtoolsMipVariable` isn't related to `operations_research::MPVariable`.

What if **OrtoolsMipVariable** has constructor taking a **MPVariable** as an argument ?

Reply:
> ```c++
> auto var_to_return = std::make_unique<OrtoolsMipVariable>(mpVar);
> return var_to_return.get();
> ```
> 
> is completely equivalent, since `var_to_return` is destructed at the end of scope.

Then we have the same problem in the **solve(...)** function :
```C++
solution_ = std::make_unique<OrtoolsMipSolution>(solution, status, objective_->Value());
return solution_.get();
```

Reply:
> > ```c++
> > auto var_to_return = std::make_unique<OrtoolsMipVariable>(mpVar);
> > return var_to_return.get();
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > is completely equivalent, since `var_to_return` is destructed at the end of scope.
> 
> Then we have the same problem in the **solve(...)** function, of which code is currently :
> 
> ```c++
> solution_ = std::make_unique<OrtoolsMipSolution>(solution, status, objective_->Value());
> return solution_.get();
> ```

No, since `solution_` is a (private) data member of class `OrtoolsLinearProblem`. As long as the instance of `OrtoolsLinearProblem`, pointer `solution_.get()` is guaranteed to remain valid (unless it is `reset`/`released`).

Reply:
ok, sorry I missed it.